### PR TITLE
User accounts

### PR DESCRIPTION
**Pull request by cabishop**
Implementation of User Accounts into the Dorset Framework. Simple CsvFileUserService added to components as an initial example of a UserService.

### CODE REVIEW

File: core/src/main/java/edu/jhuapl/dorset/Application.java [hunk 1]
```diff
@@ -71,7 +73,7 @@
      * <p>
      * Uses a null reporter that ignores new reports.
      *
-     * @param router  a router that finds the appropriate agent for a request
+     * @param router a router that finds the appropriate agent for a request
```
Comment (cash) [inline core/src/main/java/edu/jhuapl/dorset/Application.java]: The extra space between variable name and description is there on purpose. I'm trying to use this standard of JavaDoc: http://blog.joda.org/2012/11/javadoc-coding-standards.html

File: core/src/main/java/edu/jhuapl/dorset/Application.java [hunk 2]
```diff
@@ -122,35 +124,46 @@ public void addResponseFilter(ResponseFilter filter) {
...
      * Add a shutdown listener
      *
-     * @param listener  a listener that runs when shutdown() is called
+     * @param listener a listener that runs when shutdown() is called
      */
     public void addShutdownListener(ShutdownListener listener) {
         shutdownListeners.add(listener);
     }
 
+    public void setUser(User user) {
```
Comment (cash) [inline core/src/main/java/edu/jhuapl/dorset/Application.java:138]: This method needs javadoc comment - tell them when should this be used (only for single user applications)

File: core/src/main/java/edu/jhuapl/dorset/users/UserService.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package edu.jhuapl.dorset.users;
+
+/**
+ * Services the user data for a Dorset Application.
+ * <p>
+ * The users support information for the Dorset Application and for Agents.
+ */
+public interface UserService {
```
Comment (cash) [inline core/src/main/java/edu/jhuapl/dorset/users/UserService.java:27]: This interface needs work. We should enumerate the different kinds of user services and make sure the interface supports them. For example, an OAuth user service might expect one kind of map to users while an LDAP one would have a different.

Also, methods should be consistent. getCurrentUser() returns a string when I would expect it to return a user object.

File: components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    private Map<String, User> users;
+    protected String userDirPath;
+    protected String fileBaseName;
+
+    /**
+     * 
+     * FileUserService
+     * 
+     * @param userDirectoryPath Path to directory that contains User files
+     * @param fileBaseName Base string of User file names
```
Comment (cash) [inline components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java:62]: Just from the documentation, I cannot figure out what this means. Maybe add some class develop documentation that explains how this works with files/directories.

File: components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        users = new HashMap<String, User>();
+        this.userDirPath = userDirectoryPath;
+        this.fileBaseName = fileBaseName;
+    }
+
+    @Override
+    public String create(User user) throws UserException {
+        Properties prop = new Properties();
+        OutputStream output = null;
+        String userFilePath = this.userDirPath + this.fileBaseName + "-" + user.getUserName()
```
Comment (cash) [inline components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java]: What happens here if the user directory path does not include a trailing slash? It's better to convert the string to path first and then add the filename to get a new path.

Also, what happens if the user directory does not exist?

File: components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            output = new FileOutputStream(userFilePath);
+
+            Set<String> userKeys = user.getUserInformationKeys();
+            for (String key : userKeys) {
+                prop.setProperty(key, user.getUserInformation(key));
+            }
+            prop.store(output, null);
+
+        } catch (IOException e) {
+            throw new UserException("Username (" + user.getUserName() + ") already exists.");
```
Comment (cash) [inline components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java]: This message does not fit the code. I think a failure here means there was an error opening the file or writing to it (not having permission to write or a full disk are 2 possible causes).

File: components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Override
+    public String retrieve(Properties properties) throws UserException {
+        User user = new User();
+        String userName = properties.getProperty("userName");
+
+        FileInputStream in;
+        Properties props = new Properties();
+        try {
+            in = new FileInputStream(this.userDirPath + "/" + this.fileBaseName + "-" + userName
```
Comment (cash) [inline components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java]: I've seen this line of code twice which means it should be pulled out into a method. Probably a method where you pass a username in and get back a path to the file.

File: components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                            + ".properties");
+            props.load(in);
+            in.close();
+        } catch (IOException e) {
+            throw new UserException("User file (" + this.fileBaseName + "-" + userName
+                            + ".properties) does not exist.");
+        }
+
+        Enumeration<?> keySet = props.propertyNames();
```
Comment (cash) [inline components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java]: Would calling props.keySet() be cleaner - get rid of the required cast later?

File: components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java (hunk 5 above)
Comment (cabishop) [reply components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java]: @cash Do you think that I should check if the directory exists on instantiation of the constructor?

Comment (cash) [reply components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java]: Yes. Always catch errors early if possible.

File: components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java (hunk 8 above)
Comment (cabishop) [reply components/user-service/src/main/java/edu/jhuapl/dorset/fileuserservice/FileUserService.java]: @cash props.keySet() cleans the code up some but there is casting that still needs to occur over the key/value pair

        Set<Object> keySet = props.keySet();
        for (Object key : keySet) {
            user.setUserInformation((String) key, props.getProperty((String) key));
        }

**Pull request merged**